File: mvp/src/main/java/com/stanly/mvp/BaseMvpPresenter.java
package com.stanly.mvp;

public class BaseMvpPresenter<V extends MvpView> implements MvpPresenter<V> {

    private V mMvpView;

    @Override
    public void attachView(V view) {
        mMvpView = view;
    }

    @Override
    public void detachView() {
        mMvpView = null;
    }

    public boolean isViewAttached() {
        return mMvpView != null;
    }

    public V getMvpView() {
        return mMvpView;
    }

    public void checkViewAttached() {
        if (!isViewAttached()) throw new MvpViewNotAttachedException();
    }

    public static class MvpViewNotAttachedException  extends RuntimeException{
        public MvpViewNotAttachedException() {
            super("Please call Presenter.attachView(MvpView) before" +
                    " requesting data to the Presenter");
        }
    }
}

File: mvp/src/main/java/com/stanly/mvp/BaseMvpPresenter.kt
package com.stanly.mvp

class BaseMvpPresenter<V : MvpView> : MvpPresenter<V> {

    private var mvpView: V? = null


    val isViewAttached: Boolean
        get() = mvpView != null

    override fun attachView(view: V) {
        mvpView = view
    }

    override fun detachView() {
        mvpView = null
    }

    fun checkViewAttached() {
        if (!isViewAttached) throw MvpViewNotAttachedException()
    }

    class MvpViewNotAttachedException : RuntimeException("Please call Presenter.attachView(MvpView) before" + " requesting data to the Presenter")
}
